code/go/0chain.net/miner/protocol_block.go

```diff
@@ -1022,6 +1067,54 @@ func (mc *Chain) generateBlock(ctx context.Context, b *block.Block,
 	}
 
 	iterInfo.cost += cost
+	futureNonceAllowed := int64(mc.ChainConfig.TxnFutureNonce())
+
+	defer func() {
+		var (
+			deleteTxns = make([]datastore.Entity, 0, len(iterInfo.futureTxns)+len(iterInfo.pastTxns))
+			txnHashes  = make([]string, 0, len(iterInfo.futureTxns)+len(iterInfo.pastTxns))
+		)
+
+		// removes future transactions with nonce is too far away
+		if len(iterInfo.futureTxns) > 0 {
+			for _, nonceTxns := range iterInfo.futureTxns {
+				txns := nonceTxns.txns
+				if len(txns) > 0 {
+					if txns[0].Nonce-nonceTxns.nonce > futureNonceAllowed {
+						// remove all following future txns
+						for _, ft := range txns {
+							deleteTxns = append(deleteTxns, ft)
+							txnHashes = append(txnHashes, ft.Hash)
+						}
+					}
+				}
+			}
+
+			logging.Logger.Debug("remove future txns",
+				zap.Int("count", len(deleteTxns)),
+				zap.Strings("txns", txnHashes),
+				zap.Int64("future transaction limit", futureNonceAllowed))
+		}
+
+		if len(iterInfo.pastTxns) > 0 {
```

Comment:
we can't remove past transactions here. tx_pool is shared among all the blocks, not only the current one, in some situations this block state will be reverted, but transaction in the pool will be lost forever. 

we can clean up transaction based on finalized block state, but not speculative

Reply:
This is where the block generator packing the transactions, if the generator see them as past transactions, no matter how many times it re-run the block generation process, they will always be treated as past transactions unless the transaction selection algorithm is different in different tries. Also, if the transactions with the same nonce get packed, then all remaining transactions will be considered as past transactions, they wouldn't have the chance to be removed until timeout, which then users can attack the pool by submitting large number of transactions with the same nonce. They will all stuck in the the pool till timeout. 

While comparing to open the door to the attack, I think deleting those past transactions when detected may not a bad choice, as they will be removed anyway after timeout, the only drawback is we can't have them back when need to revert the block, but that's a rare case, and we can't even get the executed transactions if we want to revert a block if it's finalized. Anyway, the transaction get removed in this generator could be packed by other miners in some cases. 

Re 'clean up transactions on finalized block state', 
yes, but we would only clean up transaction packed in the finalized block, we can't afford to go through the txn pool again here to get past transactions.

Reply:
@peterlimg well, it's doubtful, in general it should be cheap to run through the in-memory pool and clean up transactions in a worker, we do not need to clean up them just now and doing it in background can be good. just new worker that get's lfb and cleanups all the past transactions